internals/overlord/logstate/syslog/syslog.go

```diff
@@ -0,0 +1,298 @@
+		options.DialTimeout = defaultDialTimeout
+	}
+}
+
+// NewClient creates a syslog client.
+func NewClient(options *ClientOptions) (*Client, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "-"
+	}
+
+	opts := *options
+	fillDefaultOptions(&opts)
+
+	u, err := url.Parse(opts.Location)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + opts.Location)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if u.Scheme != "tcp" && u.Scheme != "udp" && u.Scheme != "" {
+		return nil, fmt.Errorf("invalid syslog server address scheme %q", u.Scheme)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "tcp"
+	}
+
+	c := &Client{
+		hostname: hostname,
+		address:  u,
+		options:  &opts,
+		buffer:   make([]entryWithService, 2*opts.MaxRequestEntries),
+		labels:   make(map[string]string),
+	}
+	c.entries = c.buffer[:0]
+	return c, nil
+}
+
+// SetLabels formats the given service's labels into a structured-data section
+// for a syslog message, according to RFC5424 section 6
+func (c *Client) SetLabels(serviceName string, labels map[string]string) {
+	if len(labels) == 0 {
+		delete(c.labels, serviceName)
+		return
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", "pebble", canonicalPrivEnterpriseNum)
+
+	// Sort label keys to get deterministic output
+	keys := make([]string, 0, len(labels))
+	for key := range labels {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		fmt.Fprintf(&buf, " %s=\"", key)
+		value := labels[key]
+		// escape the value according to RFC5424 6.3.3
+		for i := 0; i < len(value); i++ {
+			// don't use "for _, c := range value" as we don't want runes
+			c := value[i]
+			if c == '"' || c == '\\' || c == ']' {
+				buf.WriteByte('\\')
+			}
+			buf.WriteByte(c)
+		}
+		buf.WriteByte('"')
+	}
+	buf.WriteByte(']')
+	c.labels[serviceName] = buf.String()
+}
+
+func (c *Client) Add(entry servicelog.Entry) error {
+	if len(c.entries) >= c.options.MaxRequestEntries {
+		// 'entries' is full - remove the first element to make room
+		// Zero the removed element to allow garbage collection
+		c.entries[0] = entryWithService{}
+		c.entries = c.entries[1:]
+	}
+
+	if len(c.entries) >= cap(c.entries) {
+		// Copy all the elements to the start of the buffer
+		copy(c.buffer, c.entries)
+
+		// Reset the view into the buffer
+		c.entries = c.buffer[:len(c.entries):len(c.buffer)]
+
+		// Zero removed elements to allow garbage collection
+		for i := len(c.entries); i < len(c.buffer); i++ {
+			c.buffer[i] = entryWithService{}
+		}
+	}
+
+	message := entry.Message
+	if len(message) > 0 && message[len(message)-1] == '\n' {
+		message = message[:len(message)-1]
+	}
+
+	c.entries = append(c.entries, entryWithService{
+		Priority:  priorityVal(FacilityUserLevelMessage, SeverityInformational),
+		Version:   1,
+		Timestamp: entry.Time.Format(time.RFC3339), // Format: 2021-05-26T12:37:01Z
+		PID:       "-",
+		MsgID:     "-",
+		Message:   message,
+		service:   entry.Service,
+	})
```

Comment:
Per above, I think this will just become:
```suggestion
	c.entries = append(c.entries, entryWithService{
		Timestamp: entry.Time.Format(time.RFC3339), // Format: 2021-05-26T12:37:01Z
		Message:   message,
		service:   entry.Service,
	})
```

Reply:
Also, we should use `time.RFC3339Nano` here for time accuracy (RFC 5424 allows fractional seconds).